File: apps/web/my-app/types/story.ts
export interface Character {
  id: string;
  slug: string;
  name: string;
  voice_profile: string;
  bulbul_speaker: string;
  avatar_url?: string;
}

export interface Choice {
  id: string;
  choice_key: string;
  text: string;
  next_node_id?: string;
}

export interface StoryNode {
  id: string;
  node_type: 'narration' | 'dialogue' | 'choice' | 'end';
  display_order: number;
  is_start: boolean;
  is_end: boolean;
  text: string;
  character?: Character;
  choices?: Choice[];
}

export interface Story {
  id: string;
  slug: string;
  title: string;
  description: string;
  language: string;
  age_range: string;
  region: string;
  moral?: string;
  duration_min: number;
  cover_image: string;
  character_count: number;
  choice_count: number;
  is_completed_translation: boolean;
  created_at: string;
}

export interface StoryDetail extends Story {
  available_languages: string[];
  characters: Character[];
  nodes: StoryNode[];
  start_node_id: string;
  updated_at: string;
}

export interface StoryFilters {
  language?: string;
  age_range?: string;
  region?: string;
  search?: string;
  page?: number;
  limit?: number;
}

export interface StoryListResponse {
  data: Story[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    total_pages: number;
    has_next: boolean;
    has_prev: boolean;
  };
}

export interface MakeChoiceRequest {
  node_id: string;
  choice_key: string;
}

export interface MakeChoiceResponse {
  success: boolean;
  choice_made: {
    node_id: string;
    choice_key: string;
    choice_text: string;
  };
  next_node: StoryNode;
  progress: {
    completion_percentage: number;
    choices_made_count: number;
  };
}

File: apps/api/app/schemas/story.py
from typing import Optional, List, Dict, Any
from uuid import UUID
from pydantic import BaseModel
from datetime import datetime


class CharacterBase(BaseModel):
    slug: str
    name: str
    voice_profile: str
    bulbul_speaker: str
    avatar_url: Optional[str] = None


class CharacterResponse(CharacterBase):
    id: UUID
    
    class Config:
        from_attributes = True


class ChoiceBase(BaseModel):
    choice_key: str
    text: str
    next_node_id: Optional[UUID] = None


class ChoiceResponse(ChoiceBase):
    id: UUID
    
    class Config:
        from_attributes = True


class StoryNodeBase(BaseModel):
    id: UUID
    node_type: str  # narration, dialogue, choice, end
    display_order: int
    is_start: bool
    is_end: bool
    text: str
    character: Optional[CharacterResponse] = None
    choices: Optional[List[ChoiceResponse]] = None


class StoryNodeResponse(StoryNodeBase):
    class Config:
        from_attributes = True


class StoryBase(BaseModel):
    slug: str
    title: str
    description: str
    language: str
    age_range: str
    region: str
    moral: Optional[str] = None
    duration_min: int
    cover_image: str


class StoryCreate(StoryBase):
    pass


class StoryResponse(StoryBase):
    id: UUID
    character_count: int = 0
    choice_count: int = 0
    is_completed_translation: bool = True
    created_at: datetime
    
    class Config:
        from_attributes = True


class StoryListResponse(BaseModel):
    data: List[StoryResponse]
    pagination: Dict[str, Any]


class StoryDetailResponse(StoryBase):
    id: UUID
    available_languages: List[str]
    characters: List[CharacterResponse]
    nodes: List[StoryNodeResponse]
    start_node_id: UUID
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class MakeChoiceRequest(BaseModel):
    node_id: UUID
    choice_key: str


class MakeChoiceResponse(BaseModel):
    success: bool
    choice_made: Dict[str, Any]
    next_node: StoryNodeResponse
    progress: Dict[str, Any]
